### PR TITLE
Mute setting

### PR DESCRIPTION
**Pull request by ghost**
For those of us who like to both test and play the game. Adds a mute checkbox in Audio -> Settings. Also displays a small string in the lobby music widget if enabled.

As always criticisms welcome; Esp feedback on layout / labeling.

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 1]
```diff
@@ -301,15 +301,35 @@ Action InitAudioPanel(Widget panel)
...
-			panel.Get<SliderWidget>("MUSIC_VOLUME").OnChange += x => Game.Sound.MusicVolume = x;
-			panel.Get<SliderWidget>("VIDEO_VOLUME").OnChange += x => Game.Sound.VideoVolume = x;
+			var muteCheckbox = panel.Get<CheckboxWidget>("MUTE_SOUND");
+			var muteCheckboxOnClick = muteCheckbox.OnClick;
+			muteCheckbox.OnClick = () =>
+			{
+				muteCheckboxOnClick();
+
+				if (ss.Mute)
+				{
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: You can/should drop the braces around the `if`:
```
if (ss.Mute)
	Game.Sound.MuteAudio();
else
{
	Game.Sound.UnmuteAudio();
	ss.Mute = false;
}

**Comment by epicelite** [comment]
Add a "mute shellmap" option.

I like to see the shellmap, but hate having to hear it.

File: OpenRA.Game/Sound/Sound.cs [hunk 2]
```diff
@@ -113,6 +113,16 @@ public void StopAudio()
...
 		}
 
+		public void MuteAudio()
+		{
+			soundEngine.Volume = 0f;
+		}
+
+		public void UnmuteAudio()
+		{
+			soundEngine.Volume = 1f;
```
Comment (Phrohdoh) [inline OpenRA.Game/Sound/Sound.cs]: You should probably be setting this to whatever the player has set, not necessarily full volume.

Comment (r34ch) [inline OpenRA.Game/Sound/Sound.cs]: @Phrohdoh Unless I am mistaken I don't think the player can set overall volume? The player can adjust the sound, music and video volume independently. This PR respects those settings.

Please correct me if I'm wrong.

Comment (Phrohdoh) [inline OpenRA.Game/Sound/Sound.cs]: You are correct, carry on.

**Comment by Mailaender** [comment]
Crashes for me when loading up the Red Alert music box after mute, unmute, mute in settings:

```
Loading mod: ra
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner Exception of type `System.InvalidOperationException`: Widget MUSIC_PANEL has no child MUTE_LABEL of type LabelWidget
     at OpenRA.Widgets.Widget.Get[T] (System.String id) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 531
   at OpenRA.Mods.Common.Widgets.Logic.MusicPlayerLogic..ctor (OpenRA.Widgets.Widget widget, OpenRA.Ruleset modRules, OpenRA.World world, System.Action onExit) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs:line 41
   at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
   at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 644
   at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 650
   at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 635
   at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 663
   at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/mcs/class/corlib/System.Reflection/ConstructorInfo.cs:line 62
   at OpenRA.ObjectCreator.CreateUsingArgs (System.Reflection.ConstructorInfo ctor, System.Collections.Generic.Dictionary`2 args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:line 101
   at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2 args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:line 66
   at OpenRA.Widgets.Widget+<PostInit>c__AnonStorey0.<>m__0 (System.String l) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 254
   at System.Linq.Enumerable+WhereSelectArrayIterator`2[TSource,TResult].MoveNext () in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/System.Core/System/Linq/Enumerable.cs:line 327
   at System.Linq.Buffer`1[TElement]..ctor (IEnumerable`1 source) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/System.Core/System/Linq/Enumerable.cs:line 2570
   at System.Linq.Enumerable.ToArray[TSource] (IEnumerable`1 source) in /home/abuild/rpmbuild/BUILD/mono-4.2.1/external/referencesource/System.Core/System/Linq/Enumerable.cs:line 830
   at OpenRA.Widgets.Widget.PostInit (OpenRA.Widgets.WidgetArgs args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 255
   at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:line 70
   at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:line 43
   at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 47
   at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic+<MainMenuLogic>c__AnonStorey0.<>m__A () in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs:line 124
   at OpenRA.Mods.Common.Widgets.ButtonWidget.<ButtonWidget>m__A (MouseInput _) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Widgets/ButtonWidget.cs:line 70
   at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (MouseInput mi) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Widgets/ButtonWidget.cs:line 158
   at OpenRA.Widgets.Widget.HandleMouseInputOuter (MouseInput mi) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 368
   at OpenRA.Widgets.Ui.HandleInput (MouseInput mi) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 87
   at OpenRA.DefaultInputHandler+<OnMouseInput>c__AnonStorey2.<>m__0 () in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:line 49
   at OpenRA.Sync.CheckSyncUnchanged[T] (OpenRA.World world, System.Func`1 fn) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Sync.cs:line 187
   at OpenRA.DefaultInputHandler.OnMouseInput (MouseInput input) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:line 49
   at OpenRA.Platforms.Default.Sdl2Input.PumpInput (IInputHandler inputHandler) in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2Input.cs:line 105
   at OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput (IInputHandler inputHandler) in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs:line 376
   at OpenRA.Renderer.EndFrame (IInputHandler inputHandler) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Renderer.cs:line 154
   at OpenRA.Game.RenderTick () in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:line 566
   at OpenRA.Game.Loop () in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:line 673
   at OpenRA.Game.Run () in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:line 691
   at OpenRA.Program.Run (System.String[] args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:line 117
   at OpenRA.Program.Main (System.String[] args) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:line 39
AL lib: (EE) alc_cleanup: 1 device not closed
```

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 3]
```diff
@@ -301,15 +301,33 @@ Action InitAudioPanel(Widget panel)
...
 			BindSliderPref(panel, "SOUND_VOLUME", ss, "SoundVolume");
 			BindSliderPref(panel, "MUSIC_VOLUME", ss, "MusicVolume");
 			BindSliderPref(panel, "VIDEO_VOLUME", ss, "VideoVolume");
 
-			// Update volume immediately
-			panel.Get<SliderWidget>("SOUND_VOLUME").OnChange += x => Game.Sound.SoundVolume = x;
-			panel.Get<SliderWidget>("MUSIC_VOLUME").OnChange += x => Game.Sound.MusicVolume = x;
-			panel.Get<SliderWidget>("VIDEO_VOLUME").OnChange += x => Game.Sound.VideoVolume = x;
+			var muteCheckbox = panel.Get<CheckboxWidget>("MUTE_SOUND");
+			var muteCheckboxOnClick = muteCheckbox.OnClick;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: I assume this wants to be
```csharp
panel.Get<CheckboxWidget>("MUTE_SOUND").OnClick = () =>
{
	[...]
}
```

instead.

**Comment by r34ch** [comment]
Fixed the crashes in Extras -> Music and now accounts for missing music.

**Comment by Mailaender** [comment]
:white_check_mark:

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 4]
```diff
@@ -301,15 +301,33 @@ Action InitAudioPanel(Widget panel)
...
+			muteCheckbox.OnClick = () =>
+			{
+				muteCheckboxOnClick();
+
+				if (ss.Mute)
+					Game.Sound.MuteAudio();
+				else
+				{
+					Game.Sound.UnmuteAudio();
+					ss.Mute = false;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Am I missing something obvious - but isn't `ss.Mute` already `false` here?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Yes, the above is probably supposed to be `(!ss.Mute)`.

**Comment by r34ch** [comment]
Remnants of a refactor. Rebased and updated.

**Comment by obrakmann** [comment]
I'm not really fond of the UI for this one. I'm all for a mute option of some description, don't get me wrong, but hiding it away in Settings/Audio is just way too inaccessible, and also kinda redundant since you can just change the volume sliders as well if you're there already. I don't have a really convincing suggestion for an alternative to make, unfortunately. Maybe a hotkey would work better.

**Comment by r34ch** [comment]
@obrakmann A hotkey would indeed make much more sense - I will add that when I get some free time. 'm' seems to be free.

My reasoning for adding this to the game was because I have the sliders adjusted at slightly different levels and for testing I hate using the sliders to mute and forgetting where they were originally.

If anyone else has ideas for positioning of this option please suggest.

**Comment by RoosterDragon** [comment]
FWIW I think having the checkbox on its own, even without a hotkey, is still useful on its own merit of not losing your slider settings.

**Comment by r34ch** [comment]
Added and assigned mute hotkey to 'M'. 

Game debug also displays a message on mute / unmute so players don't get confused if accidentally pressed.

**Comment by obrakmann** [comment]
Looks good. The hotkey doesn't work in the main menu where it probably would be the most useful, but we don't have a non-hacky way to make it work there, so :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/6fde64c10565fc3da9575c56936e6f38291d5f19)

**Pull request merged**